File: Week_03/cpp/src/combinations/_39_CombinationSum.cpp
#include <vector>
#include <iostream>

using namespace std;

class 横
{
public:
    vector<vector<int>> ret;
    vector<int> path;
    vector<vector<int>> combinationSum(vector<int> &candidates, int target)
    {
        dfs(candidates, 0, target);
        return ret;
    }
    // 通过 start 限制可以选择的元素范围去重
    // 时间 O(n^target * target) 空间 O(target)
    void dfs(vector<int> &candidates, int start, int target)
    {
        if (target < 0)
            return;
        if (target == 0)
        {
            ret.push_back(path);
            return;
        }
        for (int i = start; i < candidates.size(); i++)
        {
            int v = candidates[i];
            if (target - v < 0)
                continue;
            path.push_back(v);
            dfs(candidates, i, target - v);
            path.pop_back();
        }
    }
};

class 纵
{
public:
    vector<vector<int>> ret;
    vector<int> path;
    // 这里只会考虑在[i, n]之间未被考虑过的元素，因此不会产生重复解
    // 时间 O(2^target * target) 空间 O(target)
    vector<vector<int>> combinationSum(vector<int> &candidates, int target)
    {
        dfs(candidates, 0, target);
        return ret;
    }
    void dfs(vector<int> &candidates, int i, int target)
    {
        if (i == candidates.size())
            return;
        if (target == 0)
        {
            ret.push_back(path);
            return;
        }
        int v = candidates[i];
        if (target >= v)
        {
            path.push_back(candidates[i]);
            dfs(candidates, i, target - v);
            path.pop_back();
        }
        dfs(candidates, i + 1, target);
    }
};

File: Week_03/java/src/combinations/_39_CombinationSum.java
package combinations;

import java.util.ArrayList;
import java.util.List;

/**
 * 第一遍：2020/09/09周三 ✅
 * 第二遍：2020/09/10周三 ✅
 * 第三遍：2020/09/11周三 ✅
 * 第四遍：2020/06/28周日
 */
public class _39_CombinationSum {
    static class Solution1 {
        // 时间O(n * 2^n) 空间O(target)
        public List<List<Integer>> combinationSum(int[] candidates, int target) {
            List<Integer> path = new ArrayList<>();
            List<List<Integer>> ret = new ArrayList<>();
            dfs(candidates, 0, path, target, ret);
            return ret;
        }

        void dfs(int[] arr, int i, List<Integer> path, int target, List<List<Integer>> ret) {
            if (i == arr.length) return;
            if (target == 0) {
                ret.add(new ArrayList<>(path));
                return;
            }

            if (arr[i] <= target) {
                path.add(arr[i]);
                dfs(arr, i, path, target - arr[i], ret);
                path.remove(path.size() - 1);
            }
            dfs(arr, i + 1, path, target, ret);
        }
    }

    static class Solution2 {
        // 时间O(n * n^target) 空间O(target)
        public List<List<Integer>> combinationSum(int[] candidates, int target) {
            List<Integer> path = new ArrayList<>();
            List<List<Integer>> ret = new ArrayList<>();
            dfs(candidates, 0, path, target, ret);
            return ret;
        }

        void dfs(int[] arr, int start, List<Integer> path, int target, List<List<Integer>> ret) {
            if (target == 0) {
                ret.add(new ArrayList<>(path));
                return;
            }
            for (int i = start; i < arr.length; i++) {
                if (target < arr[i]) continue;
                path.add(arr[i]);
                dfs(arr, i, path, target - arr[i], ret);
                path.remove(path.size() - 1);
            }
        }
    }
}
